### DISCUSSION TITLE
HOWTO control DC motors

### DISCUSSION THREAD

**Discussion by TrDA-hab** [arendst/Tasmota#10387] [Show and tell]
# What it can be used for: 
the control of horizontal curtain or vertical shutters, blinds adjuster or window opener, pet feeders, opening of a water tap for watering the lawn, rotating table for subject photography, opening the ventilation flap, PTZ camera, 3D Scanner Table, linear Actuator.

# 1. MX1508 Motor Driver.
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/901.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/902.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/902-1.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210120_190559.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210120_212409.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210120_214905.jpg)

### Video of the driver's work:
 - https://youtu.be/9aGlk5LSMys

### Note:
 - you cannot control the RPM of the DC motor

### How to use it:  
 - You must add support for Shutter in my_user_config.h file.
 - Run commands in the console to test the motor (you must first configure the GPIO!):  
    `ShutterMode 1`   // Enabling "Shutter mode 1"  
    `ShutterOpenDuration 2`   // Shutter opening time = 2 seconds  
    `ShutterCloseDuration 2`  // Shutter closing time = 2 seconds  
    `Restart 1`   // Restart Tasmota  
    `ShutterOpen`   // CW motor rotation  
    `ShutterClose`  // CWW motor rotation 

Wemos Pin|GPIO|Component|Motor Signal
:-:|:-:|:-:|:-:
D1|5|Relay2|CW
D2|4|Relay1|CCW

### More information:
 - [About shutter](https://tasmota.github.io/docs/Blinds-and-Shutters/)
 - [Shutter commands](https://tasmota.github.io/docs/Commands/#shutters)

# 2. TB6612FNG Motor Driver.
### More information:
 - [About shutter](https://tasmota.github.io/docs/Blinds-and-Shutters/)
 - [Shutter commands](https://tasmota.github.io/docs/Commands/#shutters)
 - [About TB6612FNG](https://www.pololu.com/product/713)

### Video of the driver's work:
 - https://youtu.be/e9kBO-Brymo
 - https://youtu.be/3fbM7FQ-1wY

## 2.1 Using ONE motor
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/903.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/904.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210103_181349.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210123_160002.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210123_155956.jpg)

### Note:
 - you can control the RPM of the DC motor !

### How to use it:
 - You must add support for Shutter in my_user_config.h file.
 - Run commands in the console (you must first configure the GPIO!):  
    `Rule1 ON system#boot DO Backlog Power3 1; Dimmer 75 ENDON `   // Enabling the driver and setting the PWM to 75%.  
    `Rule1 1`   // Enabling Rule1  
    `WebButton4 STBY`   // To rename a web button (optional)  
 - Run commands in the console to test the motor (you must first configure the GPIO!):  
    `ShutterMode 1`   // Enabling "Shutter mode 1"  
    `Restart 1`   // Restart Tasmota  
    `ShutterOpen`   // test CW motor rotation  
    `ShutterClose`  // test CWW motor rotation 

Wemos Pin|GPIO|Component|Motor Signal
:-:|:-:|:-:|:-:
D5|14|Relay3|STBY
D6|12|Relay1|CW
D7|13|Relay2|CCW
D8|15|PWM1|PWM

## 2.2 Using DUAL motor.

![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/909.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/9010.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/909-1.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210315_224338.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210317_202159.jpg)

### Note:
 - you can control the RPM of the DC motors !

### How to use it:
 - You must add support for Shutter in my_user_config.h file.
 - Run commands in the console (you must first configure the GPIO!):  
    `SetOption68 1`   // Enable multi-channel PWM  
    `SetOption73 1`   // Enable detach buttons from relays  
    `Rule1 ON system#boot DO Backlog Power5 1; Power6 1; Power7 1; Channel6 50; Channel7 75 ENDON`   // Enable PWM to 50% and 75%.  
    `Rule1 1`   // Enable Rule1  
    `Backlog WebButton5 PWM1; WebButton6 PWM2; WebButton7 STNB`   // To rename a web button (optional)  
 - Run commands in the console to test the motor (you must first configure the GPIO!):  
    `ShutterMode 1`   // Enabling "Shutter mode 1"  
    `ShutterRelay1 1`   // for relay Relay1 and Relay2  
    `ShutterRelay2 3`   // for relay Relay3 and Relay4  
    `Restart 1`   // Restart Tasmota  
    `ShutterOpen1`   // test CW motor1 rotation  
    `ShutterClose1`  // test CWW motor1 rotation 
    `ShutterOpen2`   // test CW motor2 rotation  
    `ShutterClose2`  // test CWW motor2 rotation  
    `ShutterButton1 1 updown 0`  // Assigns button#1 to act as an "up and down" button (1x press up, 2x press down) for Shutter#1   
    `ShutterButton2 2 updown 0`  // Assigns button#2 to act as an "up and down" button (1x press up, 2x press down) for Shutter#2    

ESP-01M Pin|GPIO|Component|Motor Signal
:-:|:-:|:-:|:-:
D2|04|PWM1|PWM1 motor1
D6|12|Relay1|CW motor1
D7|13|Relay2|CCW motor1
FL|09|Button1|Up/Down#1
D4|02|PWM1|PWM2 motor2
D7|13|Relay3|CW# motor2
D8|15|Relay4|CCW motor2
FL|10|Button2|Up/Down#2
D5|14|Relay5|Stnb

# 3. Wemos motor shield.

### Video of the driver's work:
 - https://youtu.be/gah3zCJPPUc  
 - https://youtu.be/8P2wTgeavdo  

### More information:  
- https://www.wemos.cc/en/latest/d1_mini_shield/motor.html

## 3.1 Using ONE motor.

![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/905.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/906.jpg)

### How to use it:

1. You must add support for "WEMOS_MOTOR_V1" in my_user_config.h file.
2. Сonfigure the GPIO:  

Wemos Pin|GPIO|Component|Motor Signal
:-:|:-:|:-:|:-:
D1|5|I2C SCL| ---
D2|4|I2C SDA| ---
D3|0|Relay1| CW motor1
D4|2|Relay2| Stop motor1
D5|14|Relay3| CCW motor1

  **If you did everything correctly, then in the console you should see this message:**  
   `00:00:00.062 I2C: WEMOS_MOTOR_V1 found at 0x30`

3. Run commands in the console to test the motor (you must first configure the GPIO!):  
    `driver44 SETMOTOR,0,1,80`  // for CCW motor1 (80% duty - optional)  
    `driver44 SETMOTOR,0,3`  // for STOP motor1  
    `driver44 SETMOTOR,0,2,80`  // for CW motor1 (80% duty - optional)  
    `driver44 SETMOTOR,0,4`  // for standby motor1 (optional)  
    [More information!](https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_34_wemos_motor_v1.ino)  
4. Use the rules to control the motor (optional):   
  - `Rule1`  // remove comments before using the rule!   
    `ON Power1#state=1 DO Backlog Power2 0; Power3 0; driver44 SETMOTOR,0,3; delay 10; driver44 SETMOTOR,0,1,80; delay 55; driver44 SETMOTOR,0,3; Power1 0 ENDON` // CW motor1, "delay 55" controls the opening (time 55 = 5.5 sec).   
    `ON Power2#state=1 DO Backlog Power1 0; Power3 0; driver44 SETMOTOR,0,3; Power2 0 ENDON` // Stop motor1   
    `ON Power3#state=1 DO Backlog Power1 0; Power2 0; driver44 SETMOTOR,0,3; delay 10; driver44 SETMOTOR,0,2,80; delay 55; driver44 SETMOTOR,0,3; Power3 0 ENDON` //CCW motor1, "delay 55" controls the closing  (time 55 = 5.5 sec).   
  - `Rule1 1`  //  run rule1   
5. Add "logic" to the control WEB buttons (optional):   
    `Backlog WebButton1 &#8648; WebButton2 Stop1; WebButton3 &#8650`   

## 3.2 Using DUAL motor.

![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/907.jpg)
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/908.jpg)

### How to use it:

1. You must add support for "WEMOS_MOTOR_V2" in my_user_config.h file.
2. Сonfigure the GPIO:  

Wemos Pin|GPIO|Component|Motor Signal
:-:|:-:|:-:|:-:
D1|5|I2C SCL| ---
D2|4|I2C SDA| ---
D3|0|Relay1| CW motor1
D4|2|Relay2| Stop motor1
D5|14|Relay3| CCW motor1
D6|12|Relay4| CW motor2
D7|13|Relay5| Stop motor2
D8|15|Relay6| CCW motor2

  **If you did everything correctly, then in the console you should see this message:**  
   `00:00:00.062 I2C: WEMOS_MOTOR_V1 found at 0x30`

3. Run commands in the console to test the **motor1** (you must first configure the GPIO!):  
    `driver44 SETMOTOR,0,1,75`  // for CCW motor1 (75% duty - optional) 
    `driver44 SETMOTOR,0,3`  // for STOP motor1  
    `driver44 SETMOTOR,0,2,75`  // for CW motor1  (75% duty - optional)
    `driver44 SETMOTOR,0,4`  // for standby motor1 (optional)  
    [More information!](https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_34_wemos_motor_v1.ino)  
4. Run commands in the console to test the **motor2** (you must first configure the GPIO!):  
    `driver44 SETMOTOR,1,1,50`  // for CCW motor2 (50% duty - optional)  
    `driver44 SETMOTOR,1,3`  // for STOP motor2  
    `driver44 SETMOTOR,1,2,50`  // for CW motor2 (50% duty - optional)  
    `driver44 SETMOTOR,1,4`  // for standby motor2 (optional)  
    [More information!](https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_34_wemos_motor_v1.ino) 
6. Use the rules to control the **motor1** and **motor2** (optional):   
  - `Rule1`  // remove comments before using the rule!   
    `ON Power1#state=1 DO Backlog Power2 0; Power3 0; driver44 SETMOTOR,0,3; delay 10; driver44 SETMOTOR,0,1,75; delay 69; driver44 SETMOTOR,0,3; Power1 0 ENDON` //  CW motor1, "delay 69" controls the opening (time 69 = 6.9 sec.)"   
    `ON Power2#state=1 DO Backlog Power1 0; Power3 0; driver44 SETMOTOR,0,3; Power2 0 ENDON` //  Stop motor1   
    `ON Power3#state=1 DO Backlog Power1 0; Power2 0; driver44 SETMOTOR,0,3; delay 10; driver44 SETMOTOR,0,2,75; delay 69; driver44 SETMOTOR,0,3; Power3 0 ENDON` //  CCW motor1, "delay 69" controls the closing  (time 69 = 6.9 sec).   
  - `Rule1 1` // run rule1  
  
  - `Rule2`  // remove comments before using the rule!   
    `ON Power4#state=1 DO Backlog Power5 0; Power6 0; driver44 SETMOTOR,1,3; delay 10; driver44 SETMOTOR,1,1,50; delay 77; driver44 SETMOTOR,1,3; Power4 0 ENDON` //  CW motor2, "delay 77" controls the opening (time 77 = 7.7 sec).   
    `ON Power5#state=1 DO Backlog Power4 0; Power6 0; driver44 SETMOTOR,1,3; Power5 0 ENDON` //  Stop motor2    
    `ON Power6#state=1 DO Backlog Power4 0; Power5 0; driver44 SETMOTOR,1,3; delay 10; driver44 SETMOTOR,1,2,50; delay 77; driver44 SETMOTOR,1,3; Power6 0 ENDON` //  CCW motor2, "delay 77" controls the closing  (time 77 = 7.7 sec).    
  - `Rule2 1` // run rule2  
6. Add "logic" to the control WEB buttons (optional):   
    `Backlog WebButton1 &#8648; WebButton2 Stop1; WebButton3 &#8650`  
    `Backlog WebButton4 &#8634; WebButton5 Stop2; WebButton6 &#8635`  

### Additional Information:
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210103_190344.jpg)  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/906-1.jpg)  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210120_190830.jpg)  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210124_174710.jpg)  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/20210124_174415.jpg)

# Note!!! :
"WEMOS MOTOR V1" requires a firmware update. Factory installed firmware has reliability issues like I2C bus stuck. The firmware requires regular USB2TTL.
[More information!](https://hackaday.io/project/18439-motor-shield-reprogramming)  

WEMOS MOTOR Pin|GPIO|Component|USB2TTL
:-:|:-:|:-:|:-:
D1|5|RX|TX
D2|4|TX|RX
3V3|-|-|3V3
GND|-|-|GND

**!!! And be sure to short-circuit the RTS and 3V pins together on "Wemos motor shield V1".**

1. Download here for firmware "motor_shield.bin":
 - https://cdn.hackaday.io/files/18439788894176/motor_shield.bin   
 or   
 - https://github.com/TrDA-hab/Projects/blob/master/DC_motor/motor_shield.bin   

2. Download "STM32Flash" from here:
 - https://sourceforge.net/projects/stm32flash/files/

3. Run these commands in the Windows console (use your COM port number!):  
`stm32flash.exe COM4`  
`stm32flash.exe -k COM4`  
`stm32flash.exe -f -v -w motor_shield.bin COM4`  

![1](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/STM32flash-1.jpg)  
![2](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/STM32flash-2.jpg)  
![3](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/STM32flash-3.jpg)  
  
**!!! After finishing the firmware, disconnect all wires (including the 3V and RTS pin), connect the shield to the ESP device and it should work!**

## 3.3 Using Wemos motor shield HR8833 Motor Shield (now no tested).   
![1](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/hr8833_v1.0.0_1_16x16.jpg)   
https://www.wemos.cc/en/latest/d1_mini_shield/hr8833_motor.html   
https://aliexpress.com/item/1005003697301523.html   

## 3.4 Using Wemos motor shield AT8870 Motor Shield (now no tested).    
![1](https://raw.githubusercontent.com/TrDA-hab/Projects/master/DC_motor/at8870_v1.0.0_1_16x16.jpg)   
https://www.wemos.cc/en/latest/d1_mini_shield/at8870_motor.html   
https://aliexpress.com/item/1005003800396010.html   

**Best regards,   
TrDA.**

**Comment by barbudor** [comment]
Nice to see your cool solutions back here 👍 
C Новым Годом

**Reply by TrDA-hab** [reply]
> Nice to see your cool solutions back here 👍
> C Новым Годом

Thank you very much !

**Reply by Frogmore42** [reply]
Very nice write up, love all the pictures.

**Reply by TrDA-hab** [reply]
> Very nice write up, love all the pictures.

I am very pleased thank you )

**Reply by vic42** [reply]
Your doc style is very unique. I like it!

**Comment by vic42** [comment]
Driver support for Wemos Motor Shield V2 is almost complete.
Currently beta testing.

**Reply by TrDA-hab** [reply]
The "V2" driver works great. I have updated the information in this article.

**Comment by modicon1** [comment]
Hi All. Can someone point me at a forum which discusses and uses a stepper motor/A4988 solution to blinds tilting, and using arendst
/Tasmota/shutters/blinds? Thanks.

**Reply by TrDA-hab** [reply]
> Hi All. Can someone point me at a forum which discusses and uses a stepper motor/A4988 solution to blinds tilting, and using arendst
> /Tasmota/shutters/blinds? Thanks.

https://tasmota.github.io/docs/Blinds-and-Shutters/
https://github.com/arendst/Tasmota/discussions/10847

**Reply by barbudor** [reply]
If you want something more interactive, join us on Discord
https://discord.gg/Ks2Kzd4

**Reply by modicon1** [reply]
Thanks TrDA, I find that this link (https://tasmota.github.io/docs/Blinds-and-Shutters/
#10847) creates more questions for me rather than gives answers. It lacks information for stepper motors and shutters and does not provide a forum to ask questions.

I have tried Discord but have not yet gotten an answer.

Thanks to you and barbudor for your responses.

**Reply by TrDA-hab** [reply]
> Thanks TrDA, I find that this link (https://tasmota.github.io/docs/Blinds-and-Shutters/
> #10847) creates more questions for me rather than gives answers. It lacks information for stepper motors and shutters and does not provide a forum to ask questions.
> 
> I have tried Discord but have not yet gotten an answer.
> 
> Thanks to you and barbudor for your responses.

I answered you on Discord. All new issues are only on Discord!

**Reply by modicon1** [reply]
Got It. Thanks

**Comment by Obelix411** [comment]
Hi, is it possible to download the board layouts somewhere?

**Comment by dharvind** [comment]
Hi all. I was able to connect motor controller L298N and control it through a D1 mini. I have now a device (chain actuator) that comes with its own motor controller. Is there anyway to connect a D1 mini to it and control it through home automation (openhab). See attached the wiring schema of the motor controller. 
![-878407774960312236](https://user-images.githubusercontent.com/44893207/153741862-462946bb-f83a-48f0-8d84-aa9691e34a5e.jpg)

**Reply by barbudor** [reply]
What are the brown/black/blue line ?
Are you looking to totally replace this by a Tasmotized controller or have Tasmota in parrallel ?

**Reply by dharvind** [reply]
The black brown blue lines are for the rain sensor and are optional.
I was thinking in the same line as an existing linear actuator I have :
Openhab - tasmotised D1 mini - L298N - linear actuator

In this case for the chain actuator maybe:
Openhab - tasmotised D1 mini - chain controller (the photo) - chain actuator
But not sure it's possible..

**Reply by barbudor** [reply]
If there is no external buttons on which you can connect relays from a tasmota to simulate press on the buttons, th ebest is probably to remove the controller and make a full tasmota solution.

**Comment by ozett** [comment]
i tried to use the TB6612 example with one motor, 
but on tasmota 11.1.x the "dimmer"-CMD does not exitst anymore.
can you confirm this? Whats the workaround/update?

![image](https://user-images.githubusercontent.com/5937811/167810635-d5deacb2-ae5b-4d90-9cd1-cde595c1eacb.png)

**Reply by sfromis** [reply]
The Dimmer command is still very much present in current Tasmota, but only becomes active when your hardware is configured with features where Dimmer makes sense, typically PWM pins. Did you configure that?.

**Reply by ozett** [reply]
> only becomes active when your hardware is configured with features where Dimmer makes sense, typically PWM pins. Did you configure that?.

i followed this manual above here, for the dc-motors. did not recognize what command configures PMM pins and acitvates dimmer.
i used firmware from webinstaller. what am i missing?

the manual above has no further commands for configuring pwm-pins, ???

![image](https://user-images.githubusercontent.com/5937811/167999997-1d16401b-9203-4543-b5cb-1e0f96a74f79.png)

**Reply by barbudor** [reply]
You saw the screen shot, so obviously you should have picked up where in the GUI you should configure GPIO for PWM (here gpio 15 set as PWM_1)
Reading this article doesn't free you from reading general Tasmota documentation

**Reply by barbudor** [reply]
If you installed from the web installer, then you didn't followed the instructions to compile your own build with shutter support

**Reply by ozett** [reply]
> you should have picked up where in the GUI you should configure GPIO for PWM (here gpio 15 set as PWM_1)

my wemos-device is configured with GPI15 as PWM pin. 
exactly like in the screenshot of the instructions.
but dimmer-cmd failed. what else is missing?
i dont think that i have to compile all with that my_user_config.h and the actual ver11 webinstaller has is already.


> If you installed from the web installer, then you didn't followed the instructions to compile your own build with shutter support

it think the shutter support is included in the web-installer, 
i flashed with the webinstaller an than i could use SetOption80 and afterwards also change shuttermodes.
or is anything else missing? 
Only a separate PWM-Control with the "Dimmer"-cmd seemed missing. 
but maybe that completly changed from firmware 9.x to 11.x ?

setting option80 worked after flashing:
![image](https://user-images.githubusercontent.com/5937811/168008876-bb0b8d3c-9f77-4070-a77c-69ca60fb4160.png)


dont know, if changing of my_user_config.h is necessary to activate shutter-support and for using the "Dimmer-Cmd"
maybe the instructions are outdated in this regard?

![image](https://user-images.githubusercontent.com/5937811/168009434-87948d59-f978-4e39-be1f-15c6dff642a2.png)

look like shutter-support is included in the tastmota.bin from the webinstaller.

![image](https://user-images.githubusercontent.com/5937811/168011069-62e3dd37-5c78-4974-8365-cc428e63b912.png)

so to make use of the dimmer-cmd for setting the PWM manually there must is **something else to configure**....???

**Comment by ozett** [comment]
there must be something wrong with the instructions,
or something has changed from firmware 9.x to 11.x 

to use the dimmer-cmd for pwm-control with one-motor an TB6612

![image](https://user-images.githubusercontent.com/5937811/168014213-13b8509d-03b4-41b5-8414-73cfe624d0df.png)

**Comment by ozett** [comment]
Got it working.
reset the whole module to defaults,

next steps:
configured first relays 1-2, 
restart
configured relay 3 and pwm.
restart

on console:
SetOption80 1
ShutterMode 1 // Enabling "Shutter mode 1"
Important: de-activate standby of motordriver with: "Power3 on"
(or otherwise nothing happens)
now "dimmer 75" is activated...

![image](https://user-images.githubusercontent.com/5937811/168016812-42583027-c3d5-4ea5-b4b5-454e1e464654.png)

**Comment by codystoica** [comment]
Hello
I tried to connect LOLIN I2C dual motor driver shield based on HR8833 to D1 mini but it does not recognize it as a device
When I run i2cscan, a device at address 0x30 is detected, but it does not recognize it. Can you help me with this problem, what could I do?
https://github.com/wemos/LOLIN_I2C_MOTOR_Library

**Reply by barbudor** [reply]
Has of today I only see support for the V1 shield with TB6612FNG
No support for HR8833

@vic42 It looks like you were working on a driver for the shield V2
Did you completed it ?

**Reply by vic42** [reply]
**Yes**, I completed the work on the V2 motor shield and integrated it into the existing Wemos Motor Shield V1 driver code, at the time when TrDA wrote this article.

I'm not aware of a hardware revision containing a different motor driver chip. However, if the I2C interface specification has not changed it should still be supported by the V2 driver, regardless.

The genuine **Lolin Wemos Motor Shield V2 is supported**. However, you have to set the correct set of `#define`s in your build to enable this specific flavor of the driver (`USE_I2C`, `USE_WEMOS_MOTOR_V1`, `WEMOS_MOTOR_V2`).

It is a bit unfortunate that the initial driver code includes "V1" explicitly, both in its name and enabling statement.

**Comment by codystoica** [comment]
I wrote tasmota custom firmware on my D1 mini with i2c LOLIN Motor Shield and indeed it appears when boots 00:00:00.069 I2C: WEMOS_MOTOR_V1 found at 0x30.
But I honestly don't understand how I can add support for "WEMOS_MOTOR_V2" because it doesn't appear in my_user_config.h file

**Comment by vic42** [comment]
The documentation at https://tasmota.github.io/docs/Compile-your-build/ explicitly states `Do not modify my_user_config.h` and instead enter your additions and changes to `user_config_override.h`. The statement for selecting the V2 driver is

`#define WEMOS_MOTOR_V2`

Adding this line will select the V2 driver code.

You may still need the statements
```
#ifndef USE_I2C
#define USE_I2C
#endif
#ifndef USE_WEMOS_MOTOR_V1
#define USE_WEMOS_MOTOR_V1
#endif
```

**Reply by codystoica** [reply]
Thank you very much!!!!

**Reply by vic42** [reply]
Please report back what kind of hardware you tried and whether it worked, If not already documented here.
Thanks!

**Comment by codystoica** [comment]
I used Wemos D1 mini with LOLIN motor shield HR8833 with two Micro Metal Gear DC N20 Motors.
![1666719493908](https://user-images.githubusercontent.com/77228515/197843856-9afce146-1fe1-4b81-a4f2-470ad51338fc.jpg)
![1666719493933](https://user-images.githubusercontent.com/77228515/197843863-7d008db8-6558-43a9-9dfc-1652778da037.jpg)

After I compiled the firmware with the options for WEMOS_MOTOR_V2 everything works as it should

 add support for "WEMOS_MOTOR_V2" in user_config_override.h file.

#define WEMOS_MOTOR_V2

#ifndef USE_I2C
#define USE_I2C
#endif

#ifndef USE_WEMOS_MOTOR_V1
#define USE_WEMOS_MOTOR_V1
#endif

![Wemos_motor_V2](https://user-images.githubusercontent.com/77228515/197844118-ee0e6e86-3d79-4d02-a8a0-b50b65350472.png)

![Jaluzele](https://user-images.githubusercontent.com/77228515/197844209-8bdbeaa0-6551-425e-b73b-75558edcb878.png)

All driver commands have been successfully tested:
 Command format:
 * driver44 <command>,<motor>,<direction>{,<duty>}
 * command:
 *    RESET - reset a motor shield
 *    SETMOTOR - set motor state
 * motor:
 *    0 - motor A
 *    1 - motor B
 *    2 - both motors (V2 only)
 * direction:
 *    0 - short break
 *    1 - CCW
 *    2 - CW
 *    3 - stop
 *    4 - standby
 * duty (optional):
 *    0 - 100% (100% by default)

Now I will integrate the project in Domoticz to control two Venetian EU Blinds

**Reply by barbudor** [reply]
Great to see another successful Tasmota project

**Comment by vic42** [comment]
Interesting to see that it works with HR8833 shields. I didn't even have such board on hand when i wrote the driver. Seems like they finally settled on some kind of API standard.

**Comment by NicBout** [comment]
Hi there, has anyone tested the driver against a memos motor shield V3 (the AT8870 version)? Should the driver work in theory or would it need to be adapted?
Thanks!

**Reply by NicBout** [reply]
I got a few boards in and can report that the driver works with the Wemos AT8870 Motor Shield, using the firmware configuration outlined in section 3.2 of the original post.

Here is the configuration I used in my user_config_override.h to generate the tasiota firmware.

#define WEMOS_MOTOR_V2

#ifndef USE_I2C
#define USE_I2C
#endif

#ifndef USE_WEMOS_MOTOR_V1
#define USE_WEMOS_MOTOR_V1
#define WEMOS_MOTOR_V1_ADDR  0x30
#define WEMOS_MOTOR_V1_FREQ  1000
#endif

**Comment by gbot20** [comment]
Has any one got a setup for using a BTS7960 H-Bridge motor driver?
It uses 2x PWM pins to drive the motor. Pulsing PWM1 moves the motor in one direction and PWM2 in the other direction.

Even better if it was possible to do an Acceleration at the start.

Thanks

**Comment by Andre0407** [comment]
Hi,
I am trying to control 5 separate PWM outputs via MQTT using ioBroker.
If only one output is set up, all parameters are displayed. Since I don't want to control a light, the SetOption68 1 command in Tasmota has helped me a lot. However, light parameters are still sent via MQTT.
Can anyone help me so that I get the parameters "On / Off" and "Dimmer" for each PWM output?

**Reply by sfromis** [reply]
Using `SetOption68` is just another type of light; separate channels. To use PWM for other purposes, I'd recommend `SetOption15` for "basic PWM control", after which you can control each channel using the `PWM` command instead of `Dimmer`. And then, say, `PWM2 0` means powering off. Do note that you get no sliders on the Tasmota front page.

**Comment by TimelessNL** [comment]
Hi,

Interesting post, and allot of information to go through.

I do have a question about the TB6612FNG though. I'm designing a PCB that is going to use the TB6612FNG to drive 12v 2-wire LED lights, as is mostly seen in Christmas lights in which 2 stings of LED's have alternating polarities. However does anyone know if the PWM pins can be left HIGH and directly applying PWM to the IN1 and IN2 would work to pulse the corresponding output? If so I could use the [tasmota h-bridge light](https://tasmota.github.io/docs/Lights/#h-bridge) or [esphome h-bridge light](https://esphome.io/components/light/hbridge.html) integration . I came across [this](https://forum.pololu.com/t/tb6612fng-dual-motor-driver-and-pwm-signals/5697/2) post, and it seems that applying PWM to the IN1 and IN2 directly could work?

I've used the DRV8833 on 5v LEDs before, which worked great as a h-bridge light but it didn't have dedicated PWM pin per channel, but the DRV8833 is not able to run at 12v so I'm looking for a mosfet based alternative that is able to run on 12v like the TB6612FNG.

Thanks in advance 👍

**Comment by mikecay** [comment]
Hi, 
I'm trying to setup the single moter with the TB6612FNG driver.  When I sent the command` Shuttermode 1`, I get `ERROR: Shtr1 Relays are not in INTERLOCK. Pls read documentation. Shutter DISABLE. Fix and REBOOT`  Following the documentation to set interlock on does not work.  I issue the following comments, I'm just not sure if Relay 1,2,3 should be in the same group.

```
Interlock 1,2,3
Interlock ON
Shutterrelay1 1
restart 1
```

No matter what I try, whenever I issue the command` Shuttermode 1`, It always gives the same error.

Any help would be appreciated.
